File: output.cpp
/** \file output.c
 Generic output functions
 */

#include "config.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <termios.h>
#include <sys/types.h>
#include <sys/stat.h>

#ifdef HAVE_SYS_TERMIOS_H
#include <sys/termios.h>
#endif

#ifdef HAVE_SYS_IOCTL_H
#include <sys/ioctl.h>
#endif

#include <sys/time.h>
#include <unistd.h>
#include <wctype.h>

#if HAVE_NCURSES_H
#include <ncurses.h>
#else
#include <curses.h>
#endif

#if HAVE_TERMIO_H
#include <termio.h>
#endif

#if HAVE_TERM_H
#include <term.h>
#elif HAVE_NCURSES_TERM_H
#include <ncurses/term.h>
#endif

#include <signal.h>
#include <fcntl.h>
#include <dirent.h>
#include <time.h>
#include <wchar.h>


#include "fallback.h"
#include "util.h"

#include "wutil.h"
#include "expand.h"
#include "common.h"
#include "output.h"
#include "highlight.h"
#include "env.h"

/**
 Number of color names in the col array
 */
#define COLORS (sizeof(col)/sizeof(wchar_t *))

static int writeb_internal( char c );

/**
 Names of different colors. 
 */
static const wchar_t *col[]=
{
	L"black",
	L"red",
	L"green",
	L"brown",
	L"yellow",
	L"blue",
	L"magenta",
	L"purple",
	L"cyan",
	L"white"
	L"normal"
}
;

/**
 Mapping from color name (the 'col' array) to color index as used in
 ANSI color terminals, and also the fish_color_* constants defined
 in highlight.h. Non-ANSI terminals will display the wrong colors,
 since they use a different mapping.
 */
static const int col_idx[]=
{
	0, 
	1,
	2,
	3,
	3,
	4,
	5,
	5,
	6,
	7,
	FISH_COLOR_NORMAL,
};

/**
 The function used for output
 */

static int (*out)(char c) = &writeb_internal;

/**
 Name of terminal
 */
static wcstring current_term;

/* Whether term256 is supported */
static bool support_term256 = false;


void output_set_writer( int (*writer)(char) )
{
	CHECK( writer, );
	out = writer;
}

int (*output_get_writer())(char)
{
	return out;
}

static bool term256_support_is_native(void) {
    /* Return YES if we think the term256 support is "native" as opposed to forced. */
    return max_colors == 256;
}

bool output_get_supports_term256() {
    return support_term256;
}

void output_set_supports_term256(bool val) {
    support_term256 = val;
}

static unsigned char index_for_color(rgb_color_t c) {
    if (c.is_named() || ! output_get_supports_term256()) {
        return c.to_name_index();
    } else {
        return c.to_term256_index();
    }
}


static bool write_color(char *todo, unsigned char idx, bool is_fg) {
    bool result = false;
    if (idx < 16 || term256_support_is_native()) {
        /* Use tparm */
        writembs( tparm( todo, idx ) );
        result = true;
    } else {
        /* We are attempting to bypass the term here. Generate the ANSI escape sequence ourself. */
        char stridx[128];
        format_long_safe(stridx, idx);
        char buff[128] = "\x1b[";
        strcat(buff, is_fg ? "38;5;" : "48;5;");
        strcat(buff, stridx);
        strcat(buff, "m");
        
        int (*writer)(char) = output_get_writer();
        if (writer) {
            for (size_t i=0; buff[i]; i++) {
                writer(buff[i]);
            }
        }
        
        result = true;
    }
    return result;
}

static bool write_foreground_color(unsigned char idx) {
    if (set_a_foreground && set_a_foreground[0]) {
        return write_color(set_a_foreground, idx, true);
    } else if (set_foreground && set_foreground[0]) {
        return write_color(set_foreground, idx, true);
    } else {
        return false;
    }
}

static bool write_background_color(unsigned char idx) {
    if (set_a_background && set_a_background[0]) {
        return write_color(set_a_background, idx, false);
    } else if (set_background && set_background[0]) {
        return write_color(set_background, idx, false);
    } else {
        return false;
    }
}


void set_color(rgb_color_t c, rgb_color_t c2)
{
    
#if 0
    wcstring tmp = c.description();
    wcstring tmp2 = c2.description();
    printf("set_color %ls : %ls\n", tmp.c_str(), tmp2.c_str());
#endif    
    ASSERT_IS_MAIN_THREAD();
    
    const rgb_color_t normal = rgb_color_t::normal();
    static rgb_color_t last_color = rgb_color_t::normal();
	static rgb_color_t last_color2 = rgb_color_t::normal();
	static int was_bold=0;
	static int was_underline=0;
	int bg_set=0, last_bg_set=0;
    
	int is_bold = 0;
	int is_underline = 0;
    
	/*
     Test if we have at least basic support for setting fonts, colors
     and related bits - otherwise just give up...
     */
	if( !exit_attribute_mode )
	{
		return;
	}
	
    
	is_bold |= c.is_bold();
	is_bold |= c2.is_bold();
    
	is_underline |= c.is_underline();
	is_underline |= c2.is_underline();
    
	if( c.is_reset() || c2.is_reset())
	{
		c = c2 = normal;
		was_bold=0;
		was_underline=0;
        /*
         If we exit attibute mode, we must first set a color, or
         previously coloured text might lose it's
         color. Terminals are weird...
         */
        write_foreground_color(0);
		writembs( exit_attribute_mode );
		return;
	}
    
	if( was_bold && !is_bold )
	{
		/*
         Only way to exit bold mode is a reset of all attributes. 
         */
		writembs( exit_attribute_mode );
		last_color = normal;
		last_color2 = normal;
		was_bold=0;
		was_underline=0;
	}
	
	if( ! last_color2.is_normal() &&
       ! last_color2.is_reset() &&
       ! last_color2.is_ignore() )
	{
		/*
         Background was set
         */
		last_bg_set=1;
	}
    
	if( ! c2.is_normal() &&
       ! c2.is_ignore())
	{
		/*
         Background is set
         */
		bg_set=1;
        if ( c==c2 )
            c = (c2==rgb_color_t::white())?rgb_color_t::black():rgb_color_t::white();
	}
    
	if( (enter_bold_mode != 0) && (strlen(enter_bold_mode) > 0))
	{
		if(bg_set && !last_bg_set)
		{
			/*
             Background color changed and is set, so we enter bold
             mode to make reading easier. This means bold mode is
             _always_ on when the background color is set.
             */
			writembs( enter_bold_mode );
		}
		if(!bg_set && last_bg_set)
		{
			/*
             Background color changed and is no longer set, so we
             exit bold mode
             */
			writembs( exit_attribute_mode );
			was_bold=0;
			was_underline=0;
			/*
             We don't know if exit_attribute_mode resets colors, so
             we set it to something known.
             */
			if( write_foreground_color(0))
			{
				last_color=rgb_color_t::black();
			}
		}
	}
    
	if( last_color != c )
	{
		if( c.is_normal() )
		{
			write_foreground_color(0);
			writembs( exit_attribute_mode );
            
			last_color2 = rgb_color_t::normal();
			was_bold=0;
			was_underline=0;
		}
		else if( ! c.is_special() )
		{
            write_foreground_color(index_for_color(c));
		}
	}
    
	last_color = c;
    
	if( last_color2 != c2 )
	{
		if( c2.is_normal() )
		{
            write_background_color(0);
            
			writembs( exit_attribute_mode );
			if( ! last_color.is_normal())
			{
                write_foreground_color(index_for_color(last_color));
			}
			
            
			was_bold=0;
			was_underline=0;
			last_color2 = c2;
		}
		else if ( ! c2.is_special() )
		{
            write_background_color(index_for_color(c2));
			last_color2 = c2;
		}
	}
    
	/*
     Lastly, we set bold mode and underline mode correctly
     */
	if( (enter_bold_mode != 0) && (strlen(enter_bold_mode) > 0) && !bg_set )
	{
		if( is_bold && !was_bold )
		{
			if( enter_bold_mode )
			{
				writembs( tparm( enter_bold_mode ) );
			}
		}
		was_bold = is_bold;	
	}
    
	if( was_underline && !is_underline )
	{
		writembs( exit_underline_mode );		
	}
	
	if( !was_underline && is_underline )
	{
		writembs( enter_underline_mode );		
	}
	was_underline = is_underline;
	
}

/**
 Default output method, simply calls write() on stdout
 */
static int writeb_internal( char c )
{
	write_loop( 1, &c, 1 );
	return 0;
}

int writeb( tputs_arg_t b )
{
	out( b );
	return 0;
}

int writembs_internal( char *str )
{
	CHECK( str, 1 );
    
	return tputs(str,1,&writeb)==ERR?1:0;
}

int writech( wint_t ch )
{
	mbstate_t state;
	size_t i;
	char buff[MB_LEN_MAX+1];
	size_t bytes;
    
	if( ( ch >= ENCODE_DIRECT_BASE) &&
       ( ch < ENCODE_DIRECT_BASE+256) )
	{
		buff[0] = ch - ENCODE_DIRECT_BASE;
		bytes=1;
	}
	else
	{
		memset( &state, 0, sizeof(state) );
		bytes= wcrtomb( buff, ch, &state );
		
		switch( bytes )
		{
			case (size_t)(-1):
			{
				return 1;
			}
		}
	}
	
	for( i=0; i<bytes; i++ )
	{
		out( buff[i] );
	}
	return 0;
}

void writestr( const wchar_t *str )
{
	char *pos;
    
	CHECK( str, );
	
    //	while( *str )
    //		writech( *str++ );
    
	/*
     Check amount of needed space
     */
	size_t len = wcstombs( 0, str, 0 );
    
	if( len == (size_t)-1 )
	{
		debug( 1, L"Tried to print invalid wide character string" );
		return;
	}
	
	len++;
    
	/*
     Convert
     */
    char *buffer, static_buffer[256];
    if (len <= sizeof static_buffer)
        buffer = static_buffer;
    else
        buffer = new char[len];
    
	wcstombs( buffer, 
             str,
             len );
    
	/*
     Write
     */
	for( pos = buffer; *pos; pos++ )
	{
		out( *pos );
	}
    
    if (buffer != static_buffer)
        delete[] buffer;
}


void writestr_ellipsis( const wchar_t *str, int max_width )
{
	int written=0;
	int tot;
    
	CHECK( str, );
	
	tot = my_wcswidth(str);
    
	if( tot <= max_width )
	{
		writestr( str );
		return;
	}
    
	while( *str != 0 )
	{
		int w = fish_wcwidth( *str );
		if( written+w+fish_wcwidth( ellipsis_char )>max_width )
		{
			break;
		}
		written+=w;
		writech( *(str++) );
	}
    
	written += fish_wcwidth( ellipsis_char );
	writech( ellipsis_char );
    
	while( written < max_width )
	{
		written++;
		writestr( L" " );
	}
}

int write_escaped_str( const wchar_t *str, int max_len )
{
    
	wchar_t *out;
	int i;
	int len;
	int written=0;
    
	CHECK( str, 0 );
	
	out = escape( str, 1 );
	len = my_wcswidth( out );
    
	if( max_len && (max_len < len))
	{
		for( i=0; (written+fish_wcwidth(out[i]))<=(max_len-1); i++ )
		{
			writech( out[i] );
			written += fish_wcwidth( out[i] );
		}
		writech( ellipsis_char );
		written += fish_wcwidth( ellipsis_char );
		
		for( i=written; i<max_len; i++ )
		{
			writech( L' ' );
			written++;
		}
	}
	else
	{
		written = len;
		writestr( out );
	}
    
	free( out );
	return written;
}


int output_color_code( const wcstring &val, bool is_background ) {
	size_t i;
    int color=FISH_COLOR_NORMAL;
	int is_bold=0;
	int is_underline=0;
	
	if (val.empty())
		return FISH_COLOR_NORMAL;
	
    wcstring_list_t el;
	tokenize_variable_array( val, el );
	
	for(size_t j=0; j < el.size(); j++ ) {
        const wcstring &next = el.at(j);
        wcstring color_name;
        if (is_background) {
            // look for something like "--background=red"
            const wcstring prefix = L"--background=";
            if (string_prefixes_string(prefix, next)) {
                color_name = wcstring(next, prefix.size());
            }
            
        } else {
            if (next == L"--bold" || next == L"-o")
                is_bold = true;
            else if (next == L"--underline" || next == L"-u")
                is_underline = true;
            else
                color_name = next;
        }
        
        if (! color_name.empty()) {
            for( i=0; i<COLORS; i++ )
            {
                if( wcscasecmp( col[i], color_name.c_str() ) == 0 )
                {
                    color = col_idx[i];
                    break;
                }
            }
        }
		
	}
	
	return color | (is_bold?FISH_COLOR_BOLD:0) | (is_underline?FISH_COLOR_UNDERLINE:0);	
}

rgb_color_t parse_color( const wcstring &val, bool is_background ) {
    int is_bold=0;
	int is_underline=0;
    
    std::vector<rgb_color_t> candidates;
    
    wcstring_list_t el;
	tokenize_variable_array( val, el );
	
	for(size_t j=0; j < el.size(); j++ ) {
        const wcstring &next = el.at(j);
        wcstring color_name;
        if (is_background) {
            // look for something like "--background=red"
            const wcstring prefix = L"--background=";
            if (string_prefixes_string(prefix, next)) {
                color_name = wcstring(next, prefix.size());
            }
        } else {
            if (next == L"--bold" || next == L"-o")
                is_bold = true;
            else if (next == L"--underline" || next == L"-u")
                is_underline = true;
            else
                color_name = next;
        }
        
        if (! color_name.empty()) {
            rgb_color_t color = rgb_color_t(color_name);
            if (! color.is_none()) {
                candidates.push_back(color);
            }
        }
    }
    
    // Pick the best candidate
    rgb_color_t first_rgb = rgb_color_t::none(), first_named = rgb_color_t::none();
    for (size_t i=0; i < candidates.size(); i++) {
        const rgb_color_t &color = candidates.at(i);
        if (color.is_rgb() && first_rgb.is_none())
            first_rgb = color;
        if (color.is_named() && first_named.is_none())
            first_named = color;
    }
    
    // If we have both RGB and named colors, then prefer rgb if term256 is supported
    rgb_color_t result;
    if ((!first_rgb.is_none() && output_get_supports_term256()) || first_named.is_none()) {
        result = first_rgb;
    } else {
        result = first_named;
    }
    
    if (result.is_none())
        result = rgb_color_t::normal();
    
    result.set_bold(is_bold);
    result.set_underline(is_underline);
    
#if 0
    wcstring desc = result.description();
    printf("Parsed %ls from %ls (%s)\n", desc.c_str(), val.c_str(), is_background ? "background" : "foreground");
#endif
    
    return result;
}

void output_set_term( const wchar_t *term )
{
	current_term = term;
}

const wchar_t *output_get_term()
{
    return current_term.empty() ? L"<unknown>" : current_term.c_str();
}

File: src/output.rs
// Generic output functions.
use crate::color::{self, Color24, RgbColor};
use crate::common::{self, wcs2string_appending};
use crate::env::EnvVar;
use crate::screen::{is_dumb, only_grayscale};
use crate::terminal_command::{
    palette_color, rgb_color, ENTER_BOLD_MODE, ENTER_DIM_MODE, ENTER_ITALICS_MODE,
    ENTER_REVERSE_MODE, ENTER_UNDERLINE_MODE, EXIT_ATTRIBUTE_MODE, EXIT_ITALICS_MODE,
    EXIT_UNDERLINE_MODE,
};
use crate::threads::MainThread;
use crate::wchar::prelude::*;
use bitflags::bitflags;
use std::cell::{RefCell, RefMut};
use std::io::{Result, Write};
use std::ops::{Deref, DerefMut};
use std::os::fd::RawFd;
use std::sync::atomic::{AtomicU8, Ordering};

bitflags! {
    #[derive(Copy, Clone, Default)]
    pub struct ColorSupport: u8 {
        const TERM_256COLOR = 1<<0;
        const TERM_24BIT = 1<<1;
    }
}

/// Whether term256 and term24bit are supported.
static COLOR_SUPPORT: AtomicU8 = AtomicU8::new(0);

pub fn get_color_support() -> ColorSupport {
    let val = COLOR_SUPPORT.load(Ordering::Relaxed);
    ColorSupport::from_bits_truncate(val)
}

pub fn set_color_support(val: ColorSupport) {
    COLOR_SUPPORT.store(val.bits(), Ordering::Relaxed);
}

pub trait InfallibleWrite: std::io::Write + Sized {
    fn infallible_write(&mut self, buf: &[u8]) {
        let n = (self as &mut dyn std::io::Write).write(buf).unwrap();
        assert!(n == buf.len());
    }
}

impl InfallibleWrite for Vec<u8> {}

fn index_for_color(c: RgbColor) -> u8 {
    if c.is_named() || !(get_color_support().contains(ColorSupport::TERM_256COLOR)) {
        return c.to_name_index();
    }
    c.to_term256_index()
}

fn write_color(outp: &mut Outputter, foreground: bool, idx: u8) -> bool {
    if is_dumb() {
        return false;
    }
    if only_grayscale()
        && !(RgbColor {
            typ: color::Type::Named { idx },
            flags: color::Flags::DEFAULT,
        })
        .is_grayscale()
    {
        return false;
    }
    palette_color(outp, foreground, idx);
    true
}

fn write_foreground_color(outp: &mut Outputter, idx: u8) -> bool {
    write_color(outp, true, idx)
}

fn write_background_color(outp: &mut Outputter, idx: u8) -> bool {
    write_color(outp, false, idx)
}

pub struct Outputter {
    /// Storage for buffered contents.
    contents: Vec<u8>,

    /// Count of how many outstanding begin_buffering() calls there are.
    buffer_count: u32,

    /// fd to output to, or -1 for none.
    fd: RawFd,

    /// Foreground.
    last_color: RgbColor,

    /// Background.
    last_color2: RgbColor,

    was_bold: bool,
    was_underline: bool,
    was_italics: bool,
    was_dim: bool,
    was_reverse: bool,
}

impl Outputter {
    /// Construct an outputter which outputs to a given fd.
    /// If the fd is negative, the outputter will buffer its output.
    const fn new_from_fd(fd: RawFd) -> Self {
        Self {
            contents: Vec::new(),
            buffer_count: 0,
            fd,
            last_color: RgbColor::NORMAL,
            last_color2: RgbColor::NORMAL,
            was_bold: false,
            was_underline: false,
            was_italics: false,
            was_dim: false,
            was_reverse: false,
        }
    }

    /// Construct an outputter which outputs to its string buffer.
    pub fn new_buffering() -> Self {
        Self::new_from_fd(-1)
    }

    fn reset_modes(&mut self) {
        self.was_bold = false;
        self.was_underline = false;
        self.was_italics = false;
        self.was_dim = false;
        self.was_reverse = false;
    }

    fn can_flush(&self) -> bool {
        self.fd >= 0 && self.buffer_count == 0
    }

    fn maybe_flush(&mut self) {
        if self.can_flush() {
            self.flush_to(self.fd);
        }
    }

    /// Add to the buffer. Assumes we are either in a buffering state, or not backed by a file
    /// at all.
    pub fn extend(&mut self, str: &[u8]) {
        assert!(!self.can_flush());
        self.infallible_write(str);
    }

    /// Unconditionally write the color string to the output.
    /// Exported for builtin_set_color's usage only.
    pub fn write_color(&mut self, color: RgbColor, is_fg: bool) -> bool {
        let supports_term24bit = get_color_support().contains(ColorSupport::TERM_24BIT);
        if !supports_term24bit || !color.is_rgb() {
            // Indexed or non-24 bit color.
            let idx = index_for_color(color);
            if is_fg {
                return write_foreground_color(self, idx);
            } else {
                return write_background_color(self, idx);
            };
        }

        if is_dumb() {
            return false;
        }
        if only_grayscale() && color.is_grayscale() {
            return false;
        }

        // 24 bit!
        let Color24 { r, g, b } = color.to_color24();
        rgb_color(self, is_fg, r, g, b);
        true
    }

    /// Sets the fg and bg color. May be called as often as you like, since if the new color is the same
    /// as the previous, nothing will be written. Negative values for set_color will also be ignored.
    /// Since the command this function emits can potentially cause the screen to flicker, the
    /// function takes care to write as little as possible.
    ///
    /// Possible values for colors are RgbColor colors or special values like RgbColor::NORMAL
    ///
    /// In order to set the color to normal, three commands may have to be written.
    ///
    /// - First a command to set the color, such as set_a_foreground. This is needed because otherwise
    /// the previous strings colors might be removed as well.
    ///
    /// - After that we write the exit_attribute_mode command to reset all color attributes.
    ///
    /// - Lastly we may need to write set_a_background or set_a_foreground to set the other half of the
    /// color pair to what it should be.
    #[allow(clippy::if_same_then_else)]
    pub fn set_color(&mut self, mut fg: RgbColor, mut bg: RgbColor) {
        // Test if we have at least basic support for setting fonts, colors and related bits - otherwise
        // just give up...
        if is_dumb() {
            return;
        }

        const normal: RgbColor = RgbColor::NORMAL;
        let mut bg_set = false;
        let mut last_bg_set = false;
        let is_bold = fg.is_bold() || bg.is_bold();
        let is_underline = fg.is_underline() || bg.is_underline();
        let is_italics = fg.is_italics() || bg.is_italics();
        let is_dim = fg.is_dim() || bg.is_dim();
        let is_reverse = fg.is_reverse() || bg.is_reverse();

        if fg.is_reset() || bg.is_reset() {
            #[allow(unused_assignments)]
            {
                fg = normal;
                bg = normal;
            }
            self.reset_modes();
            // If we exit attribute mode, we must first set a color, or previously colored text might
            // lose its color. Terminals are weird...
            write_foreground_color(self, 0);
            self.infallible_write(EXIT_ATTRIBUTE_MODE);
            return;
        }
        if (self.was_bold && !is_bold)
            || (self.was_dim && !is_dim)
            || (self.was_reverse && !is_reverse)
        {
            // Only way to exit bold/dim/reverse mode is a reset of all attributes.
            self.infallible_write(EXIT_ATTRIBUTE_MODE);
            self.last_color = normal;
            self.last_color2 = normal;
            self.reset_modes();
        }
        if !self.last_color2.is_special() {
            // Background was set.
            // "Special" here refers to the special "normal", "reset" and "none" colors,
            // that really just disable the background.
            last_bg_set = true;
        }
        if !bg.is_special() {
            // Background is set.
            bg_set = true;
            if fg == bg {
                fg = if bg == RgbColor::WHITE {
                    RgbColor::BLACK
                } else {
                    RgbColor::WHITE
                };
            }
        }

        if bg_set && !last_bg_set {
            // Background color changed and is set, so we enter bold mode to make reading easier.
            // This means bold mode is _always_ on when the background color is set.
            self.infallible_write(ENTER_BOLD_MODE);
        }
        if !bg_set && last_bg_set {
            // Background color changed and is no longer set, so we exit bold mode.
            self.infallible_write(EXIT_ATTRIBUTE_MODE);
            self.reset_modes();
            // We don't know if exit_attribute_mode resets colors, so we set it to something known.
            if write_foreground_color(self, 0) {
                self.last_color = RgbColor::BLACK;
            }
        }

        if self.last_color != fg {
            if fg.is_normal() {
                write_foreground_color(self, 0);
                self.infallible_write(EXIT_ATTRIBUTE_MODE);

                self.last_color2 = RgbColor::NORMAL;
                self.reset_modes();
            } else if !fg.is_special() {
                self.write_color(fg, true /* foreground */);
            }
        }
        self.last_color = fg;

        if self.last_color2 != bg {
            if bg.is_normal() {
                write_background_color(self, 0);

                self.infallible_write(EXIT_ATTRIBUTE_MODE);
                if !self.last_color.is_normal() {
                    self.write_color(self.last_color, true /* foreground */);
                }
                self.reset_modes();
                self.last_color2 = bg;
            } else if !bg.is_special() {
                self.write_color(bg, false /* not foreground */);
                self.last_color2 = bg;
            }
        }

        // Lastly, we set bold, underline, italics, dim, and reverse modes correctly.
        if is_bold && !self.was_bold && !bg_set {
            self.infallible_write(ENTER_BOLD_MODE);
            self.was_bold = is_bold;
        }

        if !self.was_underline && is_underline {
            self.infallible_write(ENTER_UNDERLINE_MODE);
            self.was_underline = is_underline;
        } else if self.was_underline && !is_underline {
            self.infallible_write(EXIT_UNDERLINE_MODE);
            self.was_underline = is_underline;
        }

        if self.was_italics && !is_italics {
            self.infallible_write(EXIT_ITALICS_MODE);
            self.was_italics = is_italics;
        } else if !self.was_italics && is_italics {
            self.infallible_write(ENTER_ITALICS_MODE);
            self.was_italics = is_italics;
        }

        if is_dim && !self.was_dim {
            self.infallible_write(ENTER_DIM_MODE);
            self.was_dim = is_dim;
        }
        // N.B. there is no exit_dim_mode, it's handled by exit_attribute_mode above.

        if is_reverse && !self.was_reverse {
            // Some terms do not have a reverse mode set, so standout mode is a fallback.
            self.infallible_write(ENTER_REVERSE_MODE);
            self.was_reverse = is_reverse;
        }
    }

    /// Write a wide character to the receiver.
    pub fn writech(&mut self, ch: char) {
        self.write_wstr(wstr::from_char_slice(&[ch]));
    }

    /// Write a narrow character to the receiver.
    pub fn push(&mut self, ch: u8) {
        self.contents.push(ch);
        self.maybe_flush();
    }

    /// Write a wide string.
    pub fn write_wstr(&mut self, str: &wstr) {
        wcs2string_appending(&mut self.contents, str);
        self.maybe_flush();
    }

    /// Return the "output" contents.
    pub fn contents(&self) -> &[u8] {
        &self.contents
    }

    /// Output any buffered data to the given `fd`.
    fn flush_to(&mut self, fd: RawFd) {
        if fd >= 0 && !self.contents.is_empty() {
            let _ = common::write_loop(&fd, &self.contents);
            self.contents.clear();
        }
    }

    /// Begins buffering. Output will not be automatically flushed until a corresponding
    /// end_buffering() call.
    pub fn begin_buffering(&mut self) {
        self.buffer_count += 1;
        assert!(self.buffer_count > 0, "buffer_count overflow");
    }

    /// Balance a begin_buffering() call.
    pub fn end_buffering(&mut self) {
        assert!(self.buffer_count > 0, "buffer_count underflow");
        self.buffer_count -= 1;
        self.maybe_flush();
    }
}

/// Outputter implements Write, so it may be used as the receiver of the write! macro.
/// Only ASCII data should be written this way: do NOT assume that the receiver is UTF-8.
impl Write for Outputter {
    fn write(&mut self, buf: &[u8]) -> Result<usize> {
        self.contents.extend_from_slice(buf);
        self.maybe_flush();
        Ok(buf.len())
    }

    fn flush(&mut self) -> Result<()> {
        self.flush_to(self.fd);
        Ok(())
    }
}

impl InfallibleWrite for Outputter {}

impl Outputter {
    /// Access the outputter for stdout.
    /// This should only be used from the main thread.
    pub fn stdoutput() -> &'static RefCell<Outputter> {
        static STDOUTPUT: MainThread<RefCell<Outputter>> =
            MainThread::new(RefCell::new(Outputter::new_from_fd(libc::STDOUT_FILENO)));
        STDOUTPUT.get()
    }
}

pub struct BufferedOutputter<'a>(RefMut<'a, Outputter>);

impl<'a> BufferedOutputter<'a> {
    pub fn new(outputter: &'a RefCell<Outputter>) -> Self {
        let mut outputter = outputter.borrow_mut();
        outputter.begin_buffering();
        Self(outputter)
    }
}

impl<'a> Drop for BufferedOutputter<'a> {
    fn drop(&mut self) {
        self.0.end_buffering();
    }
}

impl Deref for BufferedOutputter<'_> {
    type Target = Outputter;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for BufferedOutputter<'_> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl<'a> Write for BufferedOutputter<'a> {
    fn write(&mut self, buf: &[u8]) -> Result<usize> {
        self.0.infallible_write(buf);
        Ok(buf.len())
    }

    fn flush(&mut self) -> Result<()> {
        self.0.flush().unwrap();
        Ok(())
    }
}
impl<'a> InfallibleWrite for BufferedOutputter<'a> {}

/// Given a list of RgbColor, pick the "best" one, as determined by the color support. Returns
/// RgbColor::NONE if empty.
pub fn best_color(candidates: &[RgbColor], support: ColorSupport) -> RgbColor {
    if candidates.is_empty() {
        return RgbColor::NONE;
    }

    let mut first_rgb = RgbColor::NONE;
    let mut first_named = RgbColor::NONE;
    for color in candidates {
        if first_rgb.is_none() && color.is_rgb() {
            first_rgb = *color;
        }
        if first_named.is_none() && color.is_named() {
            first_named = *color;
        }
    }

    // If we have both RGB and named colors, then prefer rgb if term256 is supported.
    let mut result;
    let has_term256 = support.contains(ColorSupport::TERM_256COLOR);
    if (!first_rgb.is_none() && has_term256) || first_named.is_none() {
        result = first_rgb;
    } else {
        result = first_named;
    }
    if result.is_none() {
        result = candidates[0];
    }
    result
}

/// Return the internal color code representing the specified color.
/// TODO: This code should be refactored to enable sharing with builtin_set_color.
///       In particular, the argument parsing still isn't fully capable.
pub fn parse_color(var: &EnvVar, is_background: bool) -> RgbColor {
    let mut result = parse_color_maybe_none(var, is_background);
    if result.is_none() {
        result.typ = color::Type::Normal;
    }
    result
}

pub fn parse_color_maybe_none(var: &EnvVar, is_background: bool) -> RgbColor {
    let mut is_bold = false;
    let mut is_underline = false;
    let mut is_italics = false;
    let mut is_dim = false;
    let mut is_reverse = false;

    let mut candidates: Vec<RgbColor> = Vec::new();

    let prefix = L!("--background=");

    let mut next_is_background = false;
    let mut color_name = WString::new();
    for next in var.as_list() {
        color_name.clear();
        #[allow(clippy::collapsible_else_if)]
        if is_background {
            if color_name.is_empty() && next_is_background {
                color_name = next.to_owned();
                next_is_background = false;
            } else if next.starts_with(prefix) {
                // Look for something like "--background=red".
                color_name = next.slice_from(prefix.char_count()).to_owned();
            } else if next == "--background" || next == "-b" {
                // Without argument attached the next token is the color
                // - if it's another option it's an error.
                next_is_background = true;
            } else if next == "--reverse" || next == "-r" {
                // Reverse should be meaningful in either context
                is_reverse = true;
            } else if next.starts_with("-b") {
                // Look for something like "-bred".
                // Yes, that length is hardcoded.
                color_name = next.slice_from(2).to_owned();
            }
        } else {
            if next == "--bold" || next == "-o" {
                is_bold = true;
            } else if next == "--underline" || next == "-u" {
                is_underline = true;
            } else if next == "--italics" || next == "-i" {
                is_italics = true;
            } else if next == "--dim" || next == "-d" {
                is_dim = true;
            } else if next == "--reverse" || next == "-r" {
                is_reverse = true;
            } else {
                color_name = next.clone();
            }
        }

        if !color_name.is_empty() {
            let color: Option<RgbColor> = RgbColor::from_wstr(&color_name);
            if let Some(color) = color {
                candidates.push(color);
            }
        }
    }

    let mut result = best_color(&candidates, get_color_support());
    result.set_bold(is_bold);
    result.set_underline(is_underline);
    result.set_italics(is_italics);
    result.set_dim(is_dim);
    result.set_reverse(is_reverse);
    result
}
